experimental/metal/direct_command_buffer.m

```diff
@@ -0,0 +1,1187 @@
+        *value = (int8_t)byte0;
+        return true;
+      }
+      break;
+    }
+    default:
+      break;
+  }
+  return false;
+}
+
+// Fills |value| by duplicating the given |pattern| into 4-bytes.
+static iree_status_t iree_hal_metal_duplicate_to_four_byte_value(const void* pattern,
+                                                                 size_t pattern_length,
+                                                                 uint32_t* value) {
+  switch (pattern_length) {
+    case 1: {
+      uint8_t single_byte = *(uint8_t*)pattern;
+      *value = (uint32_t)single_byte;
+      *value |= (*value << 8u);
+      *value |= (*value << 16u);
+      return iree_ok_status();
+    }
+    case 2: {
+      uint16_t two_bytes = *(uint16_t*)pattern;
+      *value = (uint32_t)two_bytes;
+      *value |= (*value << 16u);
+      return iree_ok_status();
+    }
+    case 4: {
+      *value = *(uint32_t*)pattern;
+      return iree_ok_status();
+    }
+
+    default:
+      break;
+  }
+  return iree_make_status(IREE_STATUS_INTERNAL, "fill pattern should have 1/2/4 bytes");
+}
+
+static iree_status_t iree_hal_metal_command_buffer_prepare_fill_buffer(
+    iree_hal_command_buffer_t* base_command_buffer, iree_hal_buffer_t* target_buffer,
+    iree_device_size_t target_offset, iree_device_size_t length, const void* pattern,
+    iree_host_size_t pattern_length) {
+  iree_hal_metal_command_buffer_t* command_buffer =
+      iree_hal_metal_command_buffer_cast(base_command_buffer);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  id<MTLBuffer> target_device_buffer =
+      iree_hal_metal_buffer_handle(iree_hal_buffer_allocated_buffer(target_buffer));
+  target_offset += iree_hal_buffer_byte_offset(target_buffer);
+
+  // Allocate the command segment and keep track of all necessary API data.
+  uint8_t* storage_base = NULL;
+  iree_hal_metal_command_segment_t* segment = NULL;
+  iree_status_t status = iree_arena_allocate(
+      &command_buffer->arena, sizeof(*segment) + pattern_length, (void**)&storage_base);
+
+  if (iree_status_is_ok(status)) {
+    // Copy the patttern to the end of the segment for later access.
+    uint8_t* pattern_ptr = storage_base + sizeof(*segment);
+    memcpy(pattern_ptr, (const uint8_t*)pattern, pattern_length);
+
+    // Compose and push the fill buffer segment.
+    segment = (iree_hal_metal_command_segment_t*)storage_base;
+    segment->action = IREE_HAL_METAL_COMMAND_SEGMENT_ACTION_FILL_BUFFER;
+    iree_hal_metal_command_segment_list_push_back(&command_buffer->segments, segment);
+
+    segment->fill_buffer.target_buffer = target_device_buffer;
+    segment->fill_buffer.target_offset = target_offset;
+    segment->fill_buffer.length = length;
+    segment->fill_buffer.pattern = (const void*)pattern_ptr;
+    segment->fill_buffer.pattern_length = pattern_length;
+  }
+
+  IREE_RETURN_IF_ERROR(
+      iree_hal_resource_set_insert(command_buffer->resource_set, 1, &target_buffer));
+
+  IREE_TRACE_ZONE_END(z0);
+  return status;
```

Comment:
Change to 
```c
if (iree_status_is_ok(status)) {
  status = ...
}
```
style, so the trace zone is not left open if there is an error

```suggestion
  if (iree_status_is_ok(status)) {
    status = iree_hal_resource_set_insert(command_buffer->resource_set, 1, &target_buffer));
  }      

  IREE_TRACE_ZONE_END(z0);
  return status;
```

(or use `IREE_RETURN_AND_END_ZONE_IF_ERROR`)

Reply:
Good point!